File: src/services/GitHubService.js
/**
 * GitHubService — acessa a API do GitHub para buscar repositórios e conteúdo.
 */

const GITHUB_API = 'https://api.github.com'
const STORAGE_KEY = 'resumeforge_github_token'

export function saveToken(token) {
  if (token) localStorage.setItem(STORAGE_KEY, token.trim())
  else localStorage.removeItem(STORAGE_KEY)
}

export function getToken() {
  return localStorage.getItem(STORAGE_KEY) || ''
}

export function clearToken() {
  localStorage.removeItem(STORAGE_KEY)
}

function headers() {
  const token = getToken()
  return {
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
  }
}

async function ghFetch(path) {
  const res = await fetch(`${GITHUB_API}${path}`, { headers: headers() })
  if (res.status === 401) throw new Error('Token invalido ou expirado.')
  if (res.status === 403) throw new Error('Rate limit atingido. Adicione um token nas configuracoes.')
  if (res.status === 404) throw new Error(`Usuario ou recurso nao encontrado: ${path}`)
  if (!res.ok) throw new Error(`GitHub retornou status ${res.status}`)
  return res.json()
}

export async function getUser(username) {
  const path = username ? `/users/${username}` : '/user'
  return ghFetch(path)
}

export async function getRepos(username, maxRepos = 15) {
  const path = username
    ? `/users/${username}/repos?sort=pushed&per_page=50&type=all`
    : `/user/repos?sort=pushed&per_page=50&affiliation=owner`

  const repos = await ghFetch(path)

  return repos
    .filter((r) => !(r.fork && r.stargazers_count === 0))
    .slice(0, maxRepos)
    .map((r) => ({
      name: r.name,
      fullName: r.full_name,
      description: r.description || '',
      language: r.language || '',
      stars: r.stargazers_count,
      forks: r.forks_count,
      fork: r.fork,
      isPrivate: r.private,
      url: r.html_url,
      pushedAt: r.pushed_at,
      topics: r.topics || [],
    }))
}

export async function getLanguages(fullName) {
  try {
    return await ghFetch(`/repos/${fullName}/languages`)
  } catch {
    return {}
  }
}

export async function getReadme(fullName) {
  try {
    const data = await ghFetch(`/repos/${fullName}/readme`)
    const decoded = atob(data.content.replace(/\n/g, ''))
    return decoded.slice(0, 2000)
  } catch {
    return ''
  }
}

export async function enrichRepos(repos, onProgress) {
  const enriched = []
  for (let i = 0; i < repos.length; i++) {
    const repo = repos[i]
    onProgress?.({ current: i + 1, total: repos.length, name: repo.name })
    const [languages, readme] = await Promise.all([
      getLanguages(repo.fullName),
      getReadme(repo.fullName),
    ])
    enriched.push({ ...repo, languages, readme })
  }
  return enriched
}

File: backend/services/github_service.py
import httpx
from backend import config

GITHUB_API = "https://api.github.com"

def _headers() -> dict:
    token = config.get_github_token()
    h = {"Accept": "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28"}
    if token:
        h["Authorization"] = f"Bearer {token}"
    return h

async def get_repos(username: str, max_repos: int = 15) -> list[dict]:
    async with httpx.AsyncClient(timeout=15, headers=_headers()) as client:
        r = await client.get(
            f"{GITHUB_API}/users/{username}/repos",
            params={"per_page": 100, "sort": "updated", "type": "owner"}
        )
        r.raise_for_status()
        repos = r.json()

    # Filter out forks without stars, take top by stars+recency
    filtered = [
        repo for repo in repos
        if not repo.get("fork") or repo.get("stargazers_count", 0) > 0
    ]
    filtered.sort(key=lambda r: (r.get("stargazers_count", 0), r.get("pushed_at", "")), reverse=True)
    return filtered[:max_repos]

async def get_languages(full_name: str) -> dict:
    try:
        async with httpx.AsyncClient(timeout=10, headers=_headers()) as client:
            r = await client.get(f"{GITHUB_API}/repos/{full_name}/languages")
            r.raise_for_status()
            return r.json()
    except Exception:
        return {}

async def get_readme(full_name: str) -> str:
    try:
        async with httpx.AsyncClient(timeout=10, headers=_headers()) as client:
            r = await client.get(
                f"{GITHUB_API}/repos/{full_name}/readme",
                headers={**_headers(), "Accept": "application/vnd.github.raw+json"}
            )
            if r.status_code == 404:
                return ""
            r.raise_for_status()
            return r.text[:2000]
    except Exception:
        return ""

async def enrich_repos(repos: list[dict], on_progress=None) -> list[dict]:
    enriched = []
    for i, repo in enumerate(repos):
        if on_progress:
            on_progress({"current": i + 1, "total": len(repos), "name": repo["name"]})
        langs = await get_languages(repo["full_name"])
        readme = await get_readme(repo["full_name"])
        enriched.append({
            "name": repo["name"],
            "full_name": repo["full_name"],
            "description": repo.get("description") or "",
            "stars": repo.get("stargazers_count", 0),
            "forks": repo.get("forks_count", 0),
            "language": repo.get("language") or "",
            "languages": list(langs.keys()),
            "pushed_at": repo.get("pushed_at", ""),
            "url": repo.get("html_url", ""),
            "readme_snippet": readme,
            "topics": repo.get("topics", []),
        })
    return enriched

async def get_authenticated_username() -> str | None:
    token = config.get_github_token()
    if not token:
        return None
    try:
        async with httpx.AsyncClient(timeout=10, headers=_headers()) as client:
            r = await client.get(f"{GITHUB_API}/user")
            r.raise_for_status()
            return r.json().get("login")
    except Exception:
        return None
